### PR TITLE
Return a detailed diff from Diff.

### PR DESCRIPTION
**Pull request by pgavlin**
These changes add support for detailed diffs to the Kubernetes provider.
The detailed diff is calculated by computing a JSON merge patch between
the desired and new state, recursively walking the result, and recording
the difference at each leaf property as indicated by the patch.

The particular difference that is recorded for a property depends on the
old and new values:
- If the patched value is nil, the property is recorded as deleted
- If the old value is nil, the property is recorded as added
- If the types of the old and new values differ, the property is recorded
  as updated
- If both values are maps, the maps are recursively compared on a
  per-property basis and added to the diff
- If both values are arrays, the arrays are recursively compared on a
  per-element basis and added to the diff
- If both values are primitives and the values differ, the property is
  recorded as updated
- Otherwise, no diff is recorded.

If a difference is present at a path and the path matches one of the
patterns in the database of force-new properties, the diff is amended to
indicate that the resource needs to be replaced due to the change in
this property.

### CODE REVIEW

File: pkg/provider/provider.go [hunk 1]
```diff
@@ -1231,3 +1266,178 @@ func parseLiveInputs(live, oldInputs *unstructured.Unstructured) *unstructured.U
 	metadata.AdoptOldAutonameIfUnnamed(inputs, live)
 	return inputs
 }
+
+// convertPatchToDiff converts the given JSON merge patch to a Pulumi detailed diff.
+func convertPatchToDiff(patch, oldLiveState map[string]interface{}, gvk schema.GroupVersionKind) (map[string]*pulumirpc.PropertyDiff, error) {
```
Comment (lblackstone) [inline pkg/provider/provider.go]:
```suggestion
func convertPatchToDiff(patch, oldLiveState map[string]interface{}, gvk schema.GroupVersionKind,
) (map[string]*pulumirpc.PropertyDiff, error) {
```

File: pkg/provider/provider.go [hunk 2]
```diff
@@ -1231,3 +1266,178 @@ func parseLiveInputs(live, oldInputs *unstructured.Unstructured) *unstructured.U
...
+			diffKind = pulumirpc.PropertyDiff_UPDATE_REPLACE
+		}
+	}
+
+	pathStr := ""
+	for _, v := range path {
+		switch v := v.(type) {
+		case string:
+			if strings.ContainsAny(v, `."[]`) {
+				pathStr = fmt.Sprintf(`%s["%s"]`, pathStr, strings.ReplaceAll(v, `"`, `\"`))
```
Comment (lblackstone) [inline pkg/provider/provider.go]:
```suggestion
				pathStr = fmt.Sprintf(`%s[%q]`, pathStr, strings.ReplaceAll(v, `"`, `\"`))
```

Comment (pgavlin) [reply pkg/provider/provider.go]: `%q` may add additional escape sequences that are not supported by the path syntax we use in the detailed diff (documented here: https://github.com/pulumi/pulumi/blob/master/sdk/proto/provider.proto#L137-L168).

File: pkg/provider/provider.go (hunk 1 above)
Comment (pgavlin) [reply pkg/provider/provider.go]: Why add a line break here? This line is still within the 120-character limit, right?

Comment (lblackstone) [reply pkg/provider/provider.go]: It's showing up at 142 for me.

File: pkg/metadata/labels.go [hunk 3]
```diff
@@ -43,7 +49,31 @@ func SetLabel(obj *unstructured.Unstructured, key, value string) {
...
+}
+
 // SetManagedByLabel sets the `app.kubernetes.io/managed-by` label to `pulumi`.
 func SetManagedByLabel(obj *unstructured.Unstructured) {
 	SetLabel(obj, "app.kubernetes.io/managed-by", "pulumi")
 }
+
+// HasManagedByLabel returns true if the object has the `app.kubernetes.io/managed-by` label set to `pulumi`.
+func HasManagedByLabel(obj *unstructured.Unstructured) bool {
+	val := GetLabel(obj, "app.kubernetes.io/managed-by")
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply pkg/metadata/labels.go]: Done.

File: pkg/provider/provider.go (hunk 1 above)
Comment (pgavlin) [reply pkg/provider/provider.go]: Oh, weird. I wonder why the linter isn't catching that.

**Comment by pgavlin** [comment]
Looks like the Istio test failed. @lblackstone can you help me interpret the output? It looks like the failure is unrelated, but I'm not sure.

**Comment by lblackstone** [comment]
@pgavlin Yeah, it looks like a couple of the Pods failed to start for some reason, so it's probably a flake. I restarted the job.

**Pull request merged**